DSharpPlus.Core/Rest/PollyRatelimitPolicy.cs

```diff
@@ -0,0 +1,152 @@
+    internal class PollyRatelimitPolicy : AsyncPolicy<HttpResponseMessage>
+    {
+        // keeps track of endpoint -> bucket hash mapping
+        private readonly ConcurrentDictionary<string, string> __endpoint_hash_mapping;
+
+        // stores one second, a very common thing, so we don't want to allocate it every time
+        private static readonly TimeSpan __one_second = TimeSpan.FromSeconds(1);
+
+        private const string GlobalBucketName = "DSharpPlus:Ratelimiting:global";
+
+        public PollyRatelimitPolicy()
+            => __endpoint_hash_mapping = new();
+
+
+        protected override async Task<HttpResponseMessage> ImplementationAsync
+        (
+            Func<Context, CancellationToken, Task<HttpResponseMessage>> action,
+            Context context,
+            CancellationToken cancellationToken,
+            bool continueOnCapturedContext = true
+        )
+        {
+            // ascertain a valid endpoint
+            if (!context.TryGetValue("endpoint", out object endpointObject) || endpointObject is not string endpoint)
+            {
+                throw new InvalidOperationException("No valid endpoint provided.");
+            }
+
+            //ascertain a valid cache.
+            if (!context.TryGetValue("cache", out object cacheObject) || cacheObject is not ICacheService cache)
+            {
+                throw new InvalidOperationException("No valid cache provided.");
+            }
+
+            // some endpoints are exempt from the global limit. account for them.
+            bool subjectToGlobalLimit = true;
+
+            if (context.TryGetValue("subject-to-global-limit", out object globalLimitObject)
+                && globalLimitObject is bool globallyLimited)
+            {
+                subjectToGlobalLimit = globallyLimited;
+            }
+
+            // already make our request, await it later
+            ValueTask<bool> awaitableBucket = cache.TryGetAsync(endpoint, out RatelimitBucket bucket);
+
+            // apply global ratelimits
+            // todo: handle a missing global bucket
+            if (subjectToGlobalLimit && await cache.TryGetAsync(GlobalBucketName, out RatelimitBucket globalBucket))
+            {
+                if (globalBucket is null)
+                {
+                    throw new InvalidOperationException("No global ratelimit bucket could be found.");
+                }
+
+                // store the current instant for a second here
+                DateTimeOffset instant = DateTimeOffset.UtcNow;
+
+                // if the bucket is already expired, reset it. technically we could save a bit of time by goto-ing out of the
+                // if clause entirely if this check succeeds, since we are then guaranteed the request will be allowed by the global bucket.
+                if (globalBucket.ResetTime < instant)
+                {
+                    globalBucket.ResetBucket(instant + __one_second);
+                }
```

Comment:
SRP dictates that the bucket should reset itself.

Reply:
where, specifically, is the practical benefit to this in the use-case of this piece of code?

Reply:
Good programming habits and little more. While the bucket technically allows requests if it expired, there ~could~ be unintended side effects?

Like fetching the bucket before it's needed; async tasks are hot, so you've got unwanted behavior there.

Reply:
See [this sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwFYBuY4gNwmXIGNyBecgYQAoAlPSLEsATkoA2AHQARAKYAbCAE8+eUpqEMSeMX1kQE8gCpQw86QFUEbAHIB7AO7aRJCW2GiAHFN6CuAD5KPQMjU3NLG3tnISA=) for an example.